src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -189,10 +203,16 @@ public MemoryPoolIterator ConsumingStart()
                 {
                     if (!consumed.IsDefault)
                     {
+                        var lengthConsumed = new MemoryPoolIterator(_head).GetLength(consumed);
+
                         returnStart = _head;
                         returnEnd = consumed.Block;
                         _head = consumed.Block;
                         _head.Start = consumed.Index;
+
+                        // Must call Subtract() after bytes have been freed, to avoid producer starting too early and growing
+                        // buffer beyond max length.
```

Comment:
Would it be more correct to call `Subtract()` after `ReturnBlocks()`?  My goal was to call `Subtract()` at the proper time, so the buffer count matches the actual consumed bytes as closely as possible.  If `Subtract()` is called before bytes have been returned to the pool, it's possible for the producer to start again and grow the buffer larger than it should be allowed.  Unlikely to matter much in practice, but I wanted to be as close as possible to correct.